### PR TITLE
Improve documentation on `ProgressTracker` class

### PR DESCRIPTION
**Pull request by adamkpickering**
I had a hard time understanding what `ProgressTracker` does and how it is used in the various backends. I figured that improving the comments and the signature of `ProgressTracker.action` would make it easier for the next person. I think I fully understand what's going on, but might have missed something - let me know if there are any inaccuracies.

### CODE REVIEW

File: src/k8s-engine/progressTracker.ts [hunk 1]
```diff
@@ -31,13 +62,10 @@ export default class ProgressTracker {
...
   /**
-   * Run a given action.  The currently running action with the highest priority
-   * will be displayed as the progress.
+   * Register an action.
    * @returns A promise that will be resolved when the passed-in promise resolves.
    */
-  action<T>(description: string, priority: number, promise: Promise<T>): Promise<T>;
-  action<T>(description: string, priority: number, fn: () => Promise<T>): Promise<T>;
-  action<T>(description: string, priority: number, v: Promise<T> | (() => Promise<T>)) {
+  action<T>(description: string, priority: number, v: Promise<T> | (() => Promise<T>)): Promise<T> {
```
Comment (mook-as) [inline src/k8s-engine/progressTracker.ts:68]: Please don't drop the overload declarations — that lets TypeScript-aware editors show clear overload options (with useful argument names).
image

File: src/k8s-engine/progressTracker.ts [hunk 2]
```diff
@@ -11,6 +19,29 @@ export default class ProgressTracker {
...
 
+  /**
+   * A function that notifies whatever needs to be notified of
+   * changes to the state of progress.
+   */
+  protected notify: (progress: K8s.KubernetesProgress) => void;
+
+  /**
+   * A progress object that is preferred over progress objects that
+   * correspond to actions when passing one to .notify.
```
Comment (mook-as) [inline src/k8s-engine/progressTracker.ts:30]: Might be useful to describe this as having, essentially, infinitely high priority.

File: src/k8s-engine/progressTracker.ts [hunk 3]
```diff
@@ -11,6 +19,29 @@ export default class ProgressTracker {
     this.notify = notify;
   }
 
+  /**
+   * A function that notifies whatever needs to be notified of
+   * changes to the state of progress.
```
Comment (mook-as) [inline src/k8s-engine/progressTracker.ts:24]: The wording here seems a bit repetitive.  How do you feel about something like:
> A function that will be called when there is any change in the progress.

**Review by mook-as — CHANGES_REQUESTED** [review]
This for doing this! It's always a bit hard to write documentation when you (think you) have a clear idea of what some API is meant to do, so it's great for people to expand upon that.

File: src/k8s-engine/progressTracker.ts (hunk 1 above)
Comment (adamkpickering) [reply src/k8s-engine/progressTracker.ts:68]: Ah I see. I didn't understand function overloading properly before this. I'll add these back.

**Comment by adamkpickering** [comment]
@mook-as I worked in your suggestions. Not sure if the end result is what you were thinking - let me know.

File: src/k8s-engine/progressTracker.ts [hunk 4]
```diff
@@ -1,7 +1,15 @@
 import * as K8s from './k8s';
 
 /**
- * ProgressTracker is used to track progress of multiple parallel actions.
+ * ProgressTracker is used to track the progress of multiple parallel actions.
+ * It invokes a callback that takes a progress object as input when one of those
+ * actions comes to a close. An "action" is effectively a Promise with some
```
Comment (mook-as) [inline src/k8s-engine/progressTracker.ts:6]: Technically, the numeric progress should be triggered before close, but I think this is fine.

File: src/k8s-engine/progressTracker.ts [hunk 5]
```diff
@@ -1,7 +1,15 @@
...
+ * associated metadata.
```
Comment (mook-as) [inline src/k8s-engine/progressTracker.ts:7]: More importantly, multiple actions are in flight concurrently.

**Review by mook-as — APPROVED** [review]

**Pull request merged**